File: sdk/php/src/Progress.php
<?php

declare(strict_types=1);

namespace Cordum\Cap;

use Basis\Nats\Client as NatsClient;
use Cordum\Agent\V1\BusPacket;
use Cordum\Agent\V1\JobCancel;
use Cordum\Agent\V1\JobProgress;
use Google\Protobuf\Timestamp;

/**
 * Progress and cancel payload constructors and emission helpers.
 */
final class ProgressHelper
{
    /**
     * Build a job progress payload.
     */
    public static function progressPayload(
        string $senderId,
        string $jobId,
        string $stepId,
        int $percent,
        string $message,
    ): string {
        $progress = new JobProgress();
        $progress->setJobId($jobId);
        $progress->setStepId($stepId);
        $progress->setPercent($percent);
        $progress->setMessage($message);

        $now = microtime(true);
        $timestamp = new Timestamp();
        $timestamp->setSeconds((int) $now);
        $timestamp->setNanos((int) (($now - (int) $now) * 1_000_000_000));

        $packet = new BusPacket();
        $packet->setSenderId($senderId);
        $packet->setProtocolVersion(Subjects::DEFAULT_PROTOCOL_VERSION);
        $packet->setCreatedAt($timestamp);
        $packet->setJobProgress($progress);

        return Codec::marshalDeterministic($packet);
    }

    /**
     * Build a job cancel payload.
     */
    public static function cancelPayload(
        string $senderId,
        string $jobId,
        string $reason,
        string $requestedBy,
    ): string {
        $cancel = new JobCancel();
        $cancel->setJobId($jobId);
        $cancel->setReason($reason);
        $cancel->setRequestedBy($requestedBy);

        $now = microtime(true);
        $timestamp = new Timestamp();
        $timestamp->setSeconds((int) $now);
        $timestamp->setNanos((int) (($now - (int) $now) * 1_000_000_000));

        $packet = new BusPacket();
        $packet->setSenderId($senderId);
        $packet->setProtocolVersion(Subjects::DEFAULT_PROTOCOL_VERSION);
        $packet->setCreatedAt($timestamp);
        $packet->setJobCancel($cancel);

        return Codec::marshalDeterministic($packet);
    }

    /**
     * Publish a progress payload to the progress subject.
     */
    public static function emitProgress(NatsClient $nc, string $payload): void
    {
        $nc->publish(Subjects::PROGRESS, $payload);
    }

    /**
     * Publish a cancel payload to the cancel subject.
     */
    public static function emitCancel(NatsClient $nc, string $payload): void
    {
        $nc->publish(Subjects::CANCEL, $payload);
    }
}

File: sdk/rust/src/progress.rs
use std::time::SystemTime;

use crate::codec::marshal_deterministic;
use crate::errors::CapError;
use crate::pb::{bus_packet::Payload, BusPacket, JobCancel, JobProgress};
use crate::subjects;

pub fn progress_payload(
    sender_id: &str,
    job_id: &str,
    step_id: &str,
    percent: i32,
    message: &str,
) -> Vec<u8> {
    let now = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap_or_default();

    let packet = BusPacket {
        sender_id: sender_id.into(),
        protocol_version: subjects::DEFAULT_PROTOCOL_VERSION,
        created_at: Some(prost_types::Timestamp {
            seconds: now.as_secs() as i64,
            nanos: now.subsec_nanos() as i32,
        }),
        payload: Some(Payload::JobProgress(JobProgress {
            job_id: job_id.into(),
            step_id: step_id.into(),
            percent,
            message: message.into(),
            ..Default::default()
        })),
        ..Default::default()
    };

    marshal_deterministic(&packet)
}

pub fn cancel_payload(
    sender_id: &str,
    job_id: &str,
    reason: &str,
    requested_by: &str,
) -> Vec<u8> {
    let now = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap_or_default();

    let packet = BusPacket {
        sender_id: sender_id.into(),
        protocol_version: subjects::DEFAULT_PROTOCOL_VERSION,
        created_at: Some(prost_types::Timestamp {
            seconds: now.as_secs() as i64,
            nanos: now.subsec_nanos() as i32,
        }),
        payload: Some(Payload::JobCancel(JobCancel {
            job_id: job_id.into(),
            reason: reason.into(),
            requested_by: requested_by.into(),
        })),
        ..Default::default()
    };

    marshal_deterministic(&packet)
}

pub async fn emit_progress(nc: &async_nats::Client, payload: &[u8]) -> Result<(), CapError> {
    nc.publish(subjects::SUBJECT_PROGRESS, payload.to_vec().into())
        .await
        .map_err(|e| CapError::publish_failed(&e.to_string()))
}

pub async fn emit_cancel(nc: &async_nats::Client, payload: &[u8]) -> Result<(), CapError> {
    nc.publish(subjects::SUBJECT_CANCEL, payload.to_vec().into())
        .await
        .map_err(|e| CapError::publish_failed(&e.to_string()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use prost::Message;

    #[test]
    fn progress_payload_valid() {
        let bytes = progress_payload("s-1", "j-42", "step-3", 75, "processing");
        assert!(!bytes.is_empty());

        let packet = BusPacket::decode(bytes.as_slice()).unwrap();
        assert_eq!(packet.sender_id, "s-1");

        if let Some(Payload::JobProgress(p)) = packet.payload {
            assert_eq!(p.job_id, "j-42");
            assert_eq!(p.step_id, "step-3");
            assert_eq!(p.percent, 75);
            assert_eq!(p.message, "processing");
        } else {
            panic!("expected job_progress payload");
        }
    }

    #[test]
    fn cancel_payload_valid() {
        let bytes = cancel_payload("s-1", "j-99", "timeout", "admin");
        assert!(!bytes.is_empty());

        let packet = BusPacket::decode(bytes.as_slice()).unwrap();
        if let Some(Payload::JobCancel(c)) = packet.payload {
            assert_eq!(c.job_id, "j-99");
            assert_eq!(c.reason, "timeout");
            assert_eq!(c.requested_by, "admin");
        } else {
            panic!("expected job_cancel payload");
        }
    }

    #[test]
    fn progress_zero_percent() {
        let bytes = progress_payload("s", "j", "st", 0, "");
        let packet = BusPacket::decode(bytes.as_slice()).unwrap();
        if let Some(Payload::JobProgress(p)) = packet.payload {
            assert_eq!(p.percent, 0);
        } else {
            panic!("expected job_progress payload");
        }
    }
}
